src/GraphQL.ApiTests/net6/GraphQL.approved.txt

```diff
@@ -1962,6 +1963,7 @@ namespace GraphQL.Types
         protected bool Equals(GraphQL.Types.IGraphType other) { }
         public override bool Equals(object? obj) { }
         public override int GetHashCode() { }
+        public virtual void Initialize(GraphQL.Types.ISchema schema) { }
```

Comment:
IMO it's not a responsibility of GraphType itself to track its "initialization flag" according to usage of that type in some schema. And what if I use one GraphType (singleton instance) within 10 schemas of my app? For example `Money` simple type with two fields - currency and amount?

Reply:
Graph types CANNOT be used within two schema instances.  It will ALWAYS bring memory leaks, and can lead to data corruption.  If you do so, you are doing so incorrectly.  This library does not support such a scenario.

In the example you gave, now the graph type will hold a reference to a `FloatGraphType` (or similar scalar) that is not part of the schema.  Meaning, this is a different reference than the `FloatGraphType` stored in the schema's `SchemaTypes` collection.  Hence, this a memory leak and is not the correct behavior.  And if there was a modification to the built-in `FloatGraphType` for one of the schemas, the other would behave incorrectly when referencing a field from the shared `Money` type.  Only scalars can be shared across schemas, and this is a supported scenario, and still supported with this change.

This change is necessary because of the high number of users that do not understand this concept and try to used scoped schemas incorrectly.  It almost always leads to frustration on their part and an issue that we need to answer.

Many services provide for an initialization flag or similar to prevent something from happening twice when it should not.  This is no different.  Our own `Schema` class prevents double initialization.

This PR also allows for schema-dependent initialization code, which was not possible previously.

Reply:
At least, that's my opinion.

Reply:
>  Only scalars can be shared across schemas

Is this mentioned in the documentation?

Reply:
> > Only scalars can be shared across schemas
> 
> Is this mentioned in the documentation?

No, but it could be.  We recommend that graph types are registered as transients in the documentation.  `.AddGraphTypes()` registers found graph types as transients and there is no overload/argument to register them otherwise.  This way if there are multiple schemas, or if the schema is registered as scoped, each schema will get their own instance of any shared graph types.  I think we have also tests to ensure that `SchemaTypes` will only create a single instance of each graph type required.

Reply:
Another problem, which this PR neither hurts nor solves, is that when using the `ResolveType` delegate of `UnionGraphType`, the delegate must return an instance of a graph type.  But since graph types are transients, it cannot be passed into the union graph type via DI.  (It can be, but it will be an un-initialized instance, so fields' `ResolvedType` properties will be null, causing a NRE.)  The easy solution is to manually register the referenced graph type as a singleton.  Having the graph type as a singleton prevents using multiple schemas with the same graph type, and prevents using the schema as a scoped instance -- the latter can cause the multithreading issues we have seen before.  Both of those scenarios will now throw an exception with this PR.

There are a couple solutions to this issue:
1. Pass `SchemaTypes` (or `ISchema`) to the delegate, so the proper graph type can be pulled from there.
2. Have the delegate return `Type` instances rather than `IGraphType` instances.  But this may prevent usage from dynamic schemas.  So I don't think that's an option either.
3. Have the delegate return a string representing the name of the graph type to use.  Not my preferred solution.